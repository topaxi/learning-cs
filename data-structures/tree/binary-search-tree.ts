import { BinarySearchTreeNode } from './binary-search-tree-node'

export class BinarySearchTree<T> {
  protected root: BinarySearchTreeNode<T> | null = null

  constructor(protected readonly compare = (a: any, b: any) => a - b) {}

  insert(value: T): BinarySearchTreeNode<T> {
    if (this.root === null) {
      this.root = new BinarySearchTreeNode(value, this.compare)

      return this.root
    }

    return this.root.insert(value)
  }

  includes(value: T): boolean {
    return this.root !== null && this.root.includes(value)
  }

  findMin(): T | null {
    return this.root && this.root.findMin()
  }

  findMax(): T | null {
    return this.root && this.root.findMax()
  }
}
