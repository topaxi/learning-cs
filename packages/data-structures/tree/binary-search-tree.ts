import { BinarySearchTreeNode } from './binary-search-tree-node'

export class BinarySearchTree<T> implements Iterable<T> {
  protected root: BinarySearchTreeNode<T> | null = null

  static from<U>(iterable: Iterable<U>) {
    let tree = new this<U>()

    for (let value of iterable) {
      tree.insert(value)
    }

    return tree
  }

  static of<U>(...values: U[]) {
    return this.from(values)
  }

  constructor(
    protected readonly compare: (a: T, b: T) => number = (a: any, b: any) =>
      a - b
  ) {}

  get balanceFactor(): number {
    return this.root === null ? 0 : this.root.balanceFactor
  }

  insert(value: T): BinarySearchTreeNode<T> {
    if (this.root === null) {
      this.root = new BinarySearchTreeNode(value, this.compare)

      return this.root
    }

    return this.root.insert(value)
  }

  includes(value: T): boolean {
    return this.root !== null && this.root.includes(value)
  }

  findMin(): T | null {
    return this.root && this.root.findMin()
  }

  findMax(): T | null {
    return this.root && this.root.findMax()
  }

  toJSON(): BinarySearchTreeNode<T> | null {
    return this.root
  }

  *[Symbol.iterator](): IterableIterator<T> {
    if (this.root !== null) {
      yield* this.root
    }
  }
}
