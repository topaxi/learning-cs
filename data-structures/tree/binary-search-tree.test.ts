import { expect } from 'chai'
import { BinarySearchTree } from './binary-search-tree'

describe('BinarySearchTree', () => {
  describe('#includes()', () => {
    it('should return true if value is in tree', () => {
      let tree = new BinarySearchTree()

      tree.insert(5)
      tree.insert(1)
      tree.insert(7)

      expect(tree.includes(5)).to.be.true
      expect(tree.includes(7)).to.be.true
      expect(tree.includes(1)).to.be.true
    })

    it('should return false if value is in tree', () => {
      let tree = new BinarySearchTree()

      tree.insert(5)
      tree.insert(1)
      tree.insert(7)

      expect(tree.includes(0)).to.be.false
      expect(tree.includes(2)).to.be.false
      expect(tree.includes(3)).to.be.false
      expect(tree.includes(4)).to.be.false
      expect(tree.includes(6)).to.be.false
      expect(tree.includes(8)).to.be.false
    })
  })

  describe('#findMin()', () => {
    it('should return minimum value of tree', () => {
      let tree = new BinarySearchTree()

      tree.insert(5)
      tree.insert(1)
      tree.insert(7)
      tree.insert(2)
      tree.insert(3)

      expect(tree.findMin()).to.equal(1)
    })
  })

  describe('#findMax()', () => {
    it('should return maximum value of tree', () => {
      let tree = new BinarySearchTree()

      tree.insert(5)
      tree.insert(1)
      tree.insert(7)
      tree.insert(2)
      tree.insert(3)

      expect(tree.findMax()).to.equal(7)
    })
  })
})
