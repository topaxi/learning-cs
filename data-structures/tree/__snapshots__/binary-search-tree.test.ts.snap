// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinarySearchTree is a binary tree 1`] = `
BinarySearchTree:
  5
 / \\
1   7
 \\
  2
   \\
    3
`;
